### PR TITLE
Allow AsyncSequence operators to be inlined in their construction

### PR DESCRIPTION
**Pull request by phausler**
This grants AsyncSequence a huge perf boost for trivial operator application. Locally this benchmarks initially at about 3 million events per second to afterwards over 30 trillion events per second (this is an artificial speed - it is just reflecting the fact that the pipeline now takes zero time since it is already computed due to inlines)!

Previously the compiler could not see past the initialization and so any application of operators were blocked by the initialization of the operator. Now the initialization is visible and consequently the application can be directly inlined into any code-gen for usage.

### CODE REVIEW

**Comment by phausler** [comment]
I can add benchmarks for this but it seems that https://github.com/apple/swift/pull/36908 has not yet been merged.

**Comment by phausler** [comment]
@swift-ci please smoke test

**Comment by eeckstein** [comment]
This is the PR you'll need for benchmarking: https://github.com/apple/swift/pull/36923

**Comment by phausler** [comment]
@swift-ci please smoke test

**Comment by bendjones** [comment]
LGTM the numbers we spoke about earlier today sound really impressive

**Comment by ZevEisenberg** [comment]
Can someone [ELI5](https://www.dictionary.com/e/slang/eli5/) how a computer running at a few gigahertz (billions of operations per second) could do anything more than a trillion times per second?

**Comment by phausler** [comment]
> Can someone [ELI5](https://www.dictionary.com/e/slang/eli5/) how a computer running at a few gigahertz (billions of operations per second) could do anything more than a trillion times per second?

the claim actually has to do more with the inline of the code rather than actual operations per second, what is happening is that the compiler can see through the implementation enough that the execution is now just a memcpy and the actual work is zero here. So perhaps the number of trillions is slightly in bad faith - but more so a testament to the amazing optimization this now can enjoy.

**Comment by phausler** [comment]
long/short, any per-item benchmark div-0's ;)

**Review by aschwaighofer — APPROVED** [review]
looks good to me
